core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java

```diff
@@ -426,6 +428,38 @@ public ProxiedAddresses proxiedAddresses() {
         return proxiedAddresses;
     }
 
+    private CompletableFuture<Void> initiateConnectionShutdown(InitiateConnectionShutdown event) {
+        if (!ch.isActive()) {
+            return UnmodifiableFuture.completedFuture(null);
+        }
+        final CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+        ch.closeFuture().addListener(f -> {
+            if (f.cause() == null) {
+                completableFuture.complete(null);
+            } else {
+                completableFuture.completeExceptionally(f.cause());
+            }
+        });
+        ch.pipeline().fireUserEventTriggered(event);
+        return completableFuture;
+    }
+
+    @Override
+    public CompletableFuture<Void> initiateConnectionShutdown(long drainDurationMicros) {
```

Comment:
I'm somewhat concerned about using micros for a `long` parameter because this is probably the only place we use micros instead of millis. Would it be a bad idea to accept a millis value here? A user can use `Duration` for micros precision.

Reply:
Yeah, this one might be confusing, but in this case I think your earlier argument about GC pressure applies as well. What if user needs micros precision and their server handles O(million) connections and they use this API? Constructing and destructing Duration objects for each of the connections will cause high GC pressure.

What do you think about making the method name more descriptive? E.g. `initiateConnectionShutdownWithDrainDurationMicros(long)` for custom micros, `initiateConnectionShutdownWithDrainDuration(Duration)` for custom duration, `initiateConnectionShutdown()` for default?  Sort of similar to the server builder API https://github.com/line/armeria/pull/3715#discussion_r678200489.